File: Javascript/bio/src/physic/PhysicObject.js
Class('bio.physic.PhysicObject', {
	constructor: function PhysicObject(x, y, width, height) {
		PhysicObject.base.call(this);
		
		x = x || 0;
		y = y || 0;
		width = width || 1;
		height = height || 1;

		this.location = new bio.physic.Vector(x, y);
		this.size = new bio.physic.Vector(width, height);
		this.movement = new bio.physic.Force(0, 0);
		this.view = null;

		this.weight = 0.0;
	},

	properties: [ 'weight' ],


	getMovement: function() { return this.movement; },

	getWidth: function() { return this.size.x; },
	setWidth: function(value) { this.size.x = value; },
	getHeight: function() { return this.size.y; },
	setHeight: function(value) { this.size.y = value; },

	getX: function() { return this.location.x; },
	setX: function(value) { this.location.x = value; },
	getY: function() { return this.location.y; },
	setY: function(value) { this.location.y = value; },

	getStartX: function() {
		return this.location.x - this.size.x / 2;
	},
	getStartY: function() {
		return this.location.y - this.size.y / 2;
	},

	getEndX: function() {
		return this.location.x + this.size.x / 2;
	},
	getEndY: function() {
		return this.location.y + this.size.y / 2;
	},

	getPosition: function() {
		return this.location.copy();
	},
	setPosition: function(x, y) {
		this.location.x = x;
		this.location.y = y;
	},

	getSize: function() {
		return this.size.copy();
	},
	setSize: function(x, y) {
		y = typeof y === 'undefined' ? x : y;
		this.size.x = x;
		this.size.y = y;
	},
	getArea: function() {
		return this.size.x * this.size.y;
	},

	getDirection: function() {
		return this.movement.getDirection();
	},
	setDirection: function(degree) {
		this.movement.setDirection(degree);
	},
	modifyDirection: function(degree) {
		this.movement.modifyDirection(degree);
	},

	getVelocity: function() {
		return this.movement.getStrength();
	},
	setVelocity: function(value) {
		this.movement.setStrength(value);
	},
	modifyVelocity: function(addition) {
		this.movement.modifyStrength(addition);
	},

	testCollision: function(target) {
		return (
			this.getStartX() < target.getEndX() &&
			this.getStartY() < target.getEndY() &&
			this.getEndX() > target.getStartX() &&
			this.getEndY() > target.getStartY()
		);
	},

	distance: function(target) {
		if (target instanceof bio.physic.PhysicObject)
			target = target.location;
		return this.location.diff(target).getHypotenuse();
	},

	angle: function(target) {
		if (target instanceof bio.physic.PhysicObject)
			target = target.location;
		return target.diff(this.location).getAngle();
	},

	getNextPosition: function() {
		return this.location.copy().merge(this.movement.getVector());
	},

	_move: function() {
		this.location.merge(this.movement.getVector());
	},
	move: function() {
		this._move();
		this.fireEvent('move', this, this.location.x, this.location.y);
	},

	shove: function(degrees, strength) {
		var force;
		if (degrees instanceof bio.physic.Force)
			force = degrees;
		else
			force = new bio.physic.Force(degrees, strength * (1 - this.weight));
		this.movement.merge(force);
	},

	accelerate: function(force) {
		this.movement.modifyStrength(force);
	},

	brake: function(force) {
		var vel = this.movement.getStrength() - force;
		if (vel < 0)
			vel = 0;
		this.movement.setStrength(vel);
	},

	stop: function() {
		this.movement = new bio.physic.Force(0, 0);
	},

	isStopped: function() {
		return Math.round(this.movement.getStrength() * 10) === 0;
	}
});

File: Dart/life/specs/physic/PhysicObject.dart
#library('bio.physic');
#import('Vector.dart');
#import('Force.dart');

class PhysicObject {
	//
	// FIELDS
	//
	Vector location;
	Vector size;
	Force movement;
	Force intent;
	double weight;


	//
	// CONSTRUCTORS
	//
	PysicObject([num x = 0], [num y = 0], [num width = 1], [num height = 1]) {
		location = new Vector(x, y);
		size = new Vector(width, height);
		movement = new Force(0, 0);
		intent = new Force(0, 0);
		weight = 0;
	}

	//
	// PROPERTIES
	//
	num get x() => location.x;
	    set x(num value) => location.x = value;
	num get y() => location.y;
	    set y(num value) => location.y = value;

	num get direction() => movement.direction;
	    set direction(num value) => movement.direction = value;
	num get velocity() => movement.strength;
	    set velocity(num value) => movement.strength = value;
	
	
	// Getters
	Vector get start() => location - (size / 2);
	Vector get end() => location + (size / 2);

	double get area() => size.x * size.y;


	//
	// METHODS
	//

	// Setters
	setPosition(num x, num y) {
		location.x = x;
		location.y = y;
	}
	setSize(num x, num y) {
		size.x = x;
		size.y = y;
	}

	// Compare with other objects
	num distance(target) {
		if (target is PhysicObject)
			target = target.location;
		if (!(target is Vector)
			throw new Error('[PhysicObject.distance(target)]: Target is not a vector');
		return (location - target).hypotenuse;
	}

	num angle(target) {
		if (target is PhysicObject)
			target = target.location;
		if (!(target is Vector)
			throw new Error('[PhysicObject.angle(target)]: Target is not a vector');
		return (location - target).angle;
	}

	bool testCollision(PhysicObject target) {
		var start = this.start;
		var end = this.end;
		var targetStart = target.start;
		var targetEnd = target.end;

		return (
			start.x < targetEnd.x &&
			start.y < targetEnd.y &&
			end.x > targetStart.x &&
			end.y > targetStart.y
		);
	}

	// Movement
	move() {
		location += movement;
		fireEvent('move', this, location.x, location.y);
	}
	shove(var degrees, num strength) {
		Force force;
		if (degrees is Force)
			force = degrees;
		else
			force = new Force(degrees, strength * (1 - weight));
		
		movement += force;
	}

	accelerate(num strength) {
		movement.strength += strength;
	}
	brake(num strength) {
		num result = movement.strength - strength;
		if (result < 0)
			result = 0;
		movement.strength = result;
	}

	stop() {
		movement = new Force(0, 0);
	}
	isStopped() {
		return Math.round(movement.strength * 10) == 0;
	}
}